// metaverse-api/lib/internal/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * This are the actions that travel thru the wire.
 * We should NEVER change this actions to honor the contract we are
 * creating and to maintain backwards compatibility.
 */
var Actions;
(function (Actions) {
    Actions[Actions["action"] = 0] = "action";
    Actions[Actions["groupLength"] = 1] = "groupLength";
    Actions[Actions["addAttribute"] = 2] = "addAttribute";
    Actions[Actions["modifyAttribute"] = 3] = "modifyAttribute";
    Actions[Actions["removeAttribute"] = 4] = "removeAttribute";
    Actions[Actions["relocateGroup"] = 5] = "relocateGroup";
    Actions[Actions["removeElement"] = 6] = "removeElement";
    Actions[Actions["addElement"] = 7] = "addElement";
    Actions[Actions["replaceElement"] = 8] = "replaceElement";
    Actions[Actions["route"] = 9] = "route";
    Actions[Actions["oldValue"] = 10] = "oldValue";
    Actions[Actions["newValue"] = 11] = "newValue";
    Actions[Actions["element"] = 12] = "element";
    Actions[Actions["group"] = 13] = "group";
    Actions[Actions["from"] = 14] = "from";
    Actions[Actions["to"] = 15] = "to";
    Actions[Actions["name"] = 16] = "name";
    Actions[Actions["data"] = 17] = "data";
    Actions[Actions["attrs"] = 18] = "attrs";
    Actions[Actions["tag"] = 19] = "tag";
    Actions[Actions["children"] = 20] = "children";
    Actions[Actions["value"] = 21] = "value";
})(Actions = exports.Actions || (exports.Actions = {}));
