// metaverse-rpc/lib/common/transports/WebSocket.js
export function WebSocketTransport(socket) {
    const queue = [];
    socket.addEventListener('open', function () {
        flush();
    });
    function flush() {
        if (socket.readyState === socket.OPEN) {
            queue.forEach($ => send($));
            queue.length = 0;
        }
    }
    function send(msg) {
        if (typeof msg === 'string') {
            socket.send(msg, { binary: false });
        }
        else if (msg instanceof Uint8Array || msg instanceof ArrayBuffer || msg instanceof SharedArrayBuffer) {
            ;
            socket.binaryType = 'arraybuffer';
            socket.send(msg, { binary: true });
        }
    }
    const api = {
        onConnect(handler) {
            if (socket.readyState === socket.OPEN) {
                handler();
            }
            else {
                socket.addEventListener('open', () => handler(), { once: true });
            }
        },
        onError(handler) {
            socket.addEventListener('error', (err) => handler(err.error));
        },
        onMessage(handler) {
            socket.addEventListener('message', (message) => {
                handler(message.data);
            });
        },
        sendMessage(message) {
            const toSend = message instanceof Uint8Array ? message.buffer : message;
            if (socket.readyState === socket.OPEN) {
                send(toSend);
            }
            else {
                queue.push(toSend);
            }
        },
        close() {
            socket.close();
        }
    };
    return api;
}
