// metaverse-rpc/bin/test.js
#!/usr/bin/env node
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const traceur = require('traceur');
traceur.require.makeDefault(function (filename) {
    return filename.indexOf('node_modules') === -1;
}, {
    asyncFunctions: true,
    asyncGenerators: true
});
const path_1 = require("path");
const http = require("http");
const express = require("express");
const titere = require("titere");
const fs = require("fs");
const WS = require('../test/server/_testWebSocketServer');
const keepOpen = process.argv.some($ => $ === '--keep-open');
const app = express();
const port = process.env.PORT || 3000;
const server = http.createServer(app);
WS.initializeWebSocketTester(server);
app.get('/', function (req, res) {
    res.sendFile(path_1.resolve(__dirname, '../test/index.html'));
});
console.log(path_1.resolve(__dirname, '../node_modules'));
app.use('/test', express.static(path_1.resolve(__dirname, '../test')));
app.use('/node_modules', express.static(path_1.resolve(__dirname, '../node_modules')));
server.listen(port, function (error) {
    if (error) {
        console.error(error);
        process.exit(1);
    }
    else {
        console.info('==> 🌎  Listening on port %s. Open up http://localhost:%s/ in your browser.', port, port);
        const options = {
            file: `http://localhost:${port}`,
            visible: keepOpen,
            height: 600,
            width: 800,
            timeout: 5 * 60 * 1000,
            args: ['--no-sandbox', '--disable-setuid-sandbox']
        };
        titere
            .run(options)
            .then(result => {
            if (result.coverage) {
                fs.writeFileSync('test/out/out.json', JSON.stringify(result.coverage));
            }
            else {
                console.error('Coverage data not found');
                process.exit(1);
            }
            if (!keepOpen)
                process.exit(result.result.stats.failures);
        })
            .catch((err) => {
            console.error(err.message || JSON.stringify(err));
            console.dir(err);
            if (!keepOpen)
                process.exit(1);
        });
    }
});
server.on('error', e => console.log(e));
