// metaverse-api/lib/a-frame.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const textAttributeParser_1 = require("./utils/textAttributeParser");
const Parser = require('parse5/lib/parser');
// Shorthands
function parse(html, options) {
    const parser = new Parser(options);
    return parser.parse(html);
}
function walk(node, callback) {
    if (callback(node) === false) {
        return false;
    }
    else {
        let childNode;
        let i = 0;
        if (typeof node.childNodes !== 'undefined') {
            i = 0;
            childNode = node.childNodes[i];
        }
        while (childNode !== undefined) {
            if (walk(childNode, callback) === false) {
                return false;
            }
            else {
                childNode = node.childNodes[++i];
            }
        }
    }
}
exports.walk = walk;
function findNode(root, name) {
    let foundNode = null;
    walk(root, function (node) {
        if (node.nodeName === name) {
            foundNode = node;
            return false;
        }
    });
    return foundNode;
}
exports.findNode = findNode;
function simplifyNode(root) {
    return {
        tag: root.nodeName,
        children: root.childNodes && root.childNodes.map($ => simplifyNode($)),
        attrs: root.attrs && root.attrs.reduce((prev, curr) => (Object.assign({}, prev, { [curr.name]: curr.value })), {})
    };
}
exports.simplifyNode = simplifyNode;
function parseAFrame(html) {
    const root = parse(html);
    const scene = findNode(root, 'scene');
    if (!scene) {
        throw new Error('The XML does not contain a <scene> tag');
    }
    // remove `node.tag like #text` recursively
    walk(scene, function (node) {
        node.childNodes = node.childNodes.filter($ => !$.nodeName.startsWith('#'));
    });
    return textAttributeParser_1.processSimplifiedNode(simplifyNode(scene));
}
exports.parseAFrame = parseAFrame;
